File: core/Sources/RibbonCore/Translations.swift
import Foundation

// Translations. The two launch translations are public domain and ship in
// the app, whole. Licensed translations (NKJV first, two more undecided)
// arrive through API.Bible — build book horizon §15 / open question §16.8,
// decided: they stream through the room's own proxy and cache only the
// book being read, because their licenses forbid shipping the text.
//
// Translation is a personal setting, not a room setting (§2.6): notes pin
// to verse addresses, so a note lands on the same verse whichever
// translation each person reads.

/// A translation's identity — a stable, lowercase key. Extensible: stored
/// state and the database carry the raw string, so adding a translation
/// never migrates anything.
public struct TranslationID: RawRepresentable, Hashable, Sendable {
    public let rawValue: String

    public init(rawValue: String) {
        self.rawValue = rawValue
    }

    public static let bsb = TranslationID(rawValue: "bsb")
    public static let web = TranslationID(rawValue: "web")
    public static let nkjv = TranslationID(rawValue: "nkjv")
}

extension TranslationID: Codable {
    public init(from decoder: Decoder) throws {
        rawValue = try decoder.singleValueContainer().decode(String.self)
    }

    public func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        try container.encode(rawValue)
    }
}

public extension TranslationID {
    /// Name as it appears in S20: "Berean Standard", "World English".
    var displayName: String {
        TranslationRegistry.translation(for: self)?.displayName ?? rawValue.uppercased()
    }

    var fullName: String {
        TranslationRegistry.translation(for: self)?.fullName ?? rawValue.uppercased()
    }
}

/// Everything the app knows about one translation.
public struct Translation: Hashable, Identifiable, Sendable {
    public enum Source: Hashable, Sendable {
        /// Public domain, converted by tools/usfx_to_json.py, shipped in
        /// the bundle, whole. Offline is first-class.
        case bundled
        /// Licensed, served by API.Bible through the backend proxy (the
        /// key never ships in the app). `bibleID` is API.Bible's id for
        /// the exact edition; empty until the license lands.
        case apiBible(bibleID: String)
    }

    public var id: TranslationID
    public var displayName: String
    public var fullName: String
    public var source: Source
    /// Whether the text data carries words-of-Jesus markup (S20's
    /// red-letter setting).
    public var redLetter: Bool

    public init(id: TranslationID, displayName: String, fullName: String, source: Source, redLetter: Bool) {
        self.id = id
        self.displayName = displayName
        self.fullName = fullName
        self.source = source
        self.redLetter = redLetter
    }

    public var isBundled: Bool {
        if case .bundled = source { return true }
        return false
    }

    /// The proxy can only serve an edition that has been named.
    public var isConfigured: Bool {
        switch source {
        case .bundled:
            return true
        case .apiBible(let bibleID):
            return !bibleID.isEmpty
        }
    }
}

public enum TranslationRegistry {
    public static let bsb = Translation(
        id: .bsb, displayName: "Berean Standard", fullName: "Berean Standard Bible",
        source: .bundled, redLetter: false)

    public static let web = Translation(
        id: .web, displayName: "World English", fullName: "World English Bible",
        source: .bundled, redLetter: true)

    /// NKJV via API.Bible. The bibleID is filled in when Thomas Nelson's
    /// license is approved on the API.Bible account; until then the
    /// translation is registered but unavailable, and never shown.
    public static let nkjv = Translation(
        id: .nkjv, displayName: "New King James", fullName: "New King James Version",
        source: .apiBible(bibleID: ""), redLetter: false)

    /// Shipped in the app, whole.
    public static let bundled: [Translation] = [bsb, web]

    /// Licensed, streamed. Two more slots are planned and undecided —
    /// they join this list the day they are chosen.
    public static let licensed: [Translation] = [nkjv]

    public static let all: [Translation] = bundled + licensed

    public static func translation(for id: TranslationID) -> Translation? {
        all.first { $0.id == id }
    }

    public static func isBundled(_ id: TranslationID) -> Bool {
        translation(for: id)?.isBundled ?? false
    }
}

File: android/core/src/main/kotlin/app/readribbon/core/Translations.kt
package app.readribbon.core

import kotlinx.serialization.KSerializer
import kotlinx.serialization.Serializable
import kotlinx.serialization.descriptors.PrimitiveKind
import kotlinx.serialization.descriptors.PrimitiveSerialDescriptor
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder

// Translations. The two launch translations are public domain and ship in
// the app, whole. Licensed translations (NKJV first, two more undecided)
// arrive through API.Bible — build book horizon §15 / open question §16.8,
// decided: they stream through the room's own proxy and cache only the
// book being read, because their licenses forbid shipping the text.
//
// Translation is a personal setting, not a room setting (§2.6): notes pin
// to verse addresses, so a note lands on the same verse whichever
// translation each person reads.

/**
 * A translation's identity — a stable, lowercase key. Extensible: stored
 * state and the database carry the raw string, so adding a translation
 * never migrates anything.
 */
@Serializable(with = TranslationIDSerializer::class)
data class TranslationID(val rawValue: String) {
    companion object {
        val bsb = TranslationID(rawValue = "bsb")
        val web = TranslationID(rawValue = "web")
        val nkjv = TranslationID(rawValue = "nkjv")
        val niv = TranslationID(rawValue = "niv")
        val nasb = TranslationID(rawValue = "nasb")
    }
}

/**
 * Swift hand-writes TranslationID's Codable conformance over a single-value
 * container, so the id travels as a bare JSON string rather than as an
 * object with a `rawValue` field. The two apps share one backend and one
 * JSON, so the wire form must not drift: this serializer is that same
 * single-value encoding.
 */
object TranslationIDSerializer : KSerializer<TranslationID> {
    override val descriptor: SerialDescriptor =
        PrimitiveSerialDescriptor("TranslationID", PrimitiveKind.STRING)

    override fun deserialize(decoder: Decoder): TranslationID =
        TranslationID(rawValue = decoder.decodeString())

    override fun serialize(encoder: Encoder, value: TranslationID) {
        encoder.encodeString(value.rawValue)
    }
}

/** Name as it appears in S20: "Berean Standard", "World English". */
val TranslationID.displayName: String
    get() = TranslationRegistry.translation(this)?.displayName ?: rawValue.uppercase()

val TranslationID.fullName: String
    get() = TranslationRegistry.translation(this)?.fullName ?: rawValue.uppercase()

/** Everything the app knows about one translation. */
data class Translation(
    val id: TranslationID,
    val displayName: String,
    val fullName: String,
    val source: Source,
    /**
     * Whether the text data carries words-of-Jesus markup (S20's
     * red-letter setting).
     */
    val redLetter: Boolean
) {
    sealed interface Source {
        /**
         * Public domain, converted by tools/usfx_to_json.py, shipped in
         * the bundle, whole. Offline is first-class.
         */
        data object bundled : Source

        /**
         * Licensed, served by API.Bible through the backend proxy (the
         * key never ships in the app). `bibleID` is API.Bible's id for
         * the exact edition; empty until the license lands.
         */
        data class apiBible(val bibleID: String) : Source
    }

    val isBundled: Boolean
        get() = source is Source.bundled

    /** The proxy can only serve an edition that has been named. */
    val isConfigured: Boolean
        get() = when (val source = source) {
            is Source.bundled -> true
            is Source.apiBible -> source.bibleID.isNotEmpty()
        }
}

object TranslationRegistry {
    val bsb = Translation(
        id = TranslationID.bsb, displayName = "Berean Standard", fullName = "Berean Standard Bible",
        source = Translation.Source.bundled, redLetter = false)

    val web = Translation(
        id = TranslationID.web, displayName = "World English", fullName = "World English Bible",
        source = Translation.Source.bundled, redLetter = true)

    // The three licensed editions, live on the room's API.Bible account
    // (Open Book plan) and served through the bible-proxy. The bibleIDs
    // are catalog identifiers, not secrets. All three carry words-of-Jesus
    // markup.
    val nkjv = Translation(
        id = TranslationID.nkjv, displayName = "New King James", fullName = "New King James Version",
        source = Translation.Source.apiBible(bibleID = "63097d2a0a2f7db3-01"), redLetter = true)

    val niv = Translation(
        id = TranslationID.niv, displayName = "New International", fullName = "New International Version (2011)",
        source = Translation.Source.apiBible(bibleID = "78a9f6124f344018-01"), redLetter = true)

    val nasb = Translation(
        id = TranslationID.nasb, displayName = "New American Standard", fullName = "New American Standard Bible (1995)",
        source = Translation.Source.apiBible(bibleID = "b8ee27bcd1cae43a-01"), redLetter = true)

    /** Shipped in the app, whole. */
    val bundled: List<Translation> = listOf(bsb, web)

    /** Licensed, streamed. */
    val licensed: List<Translation> = listOf(nkjv, niv, nasb)

    val all: List<Translation> = bundled + licensed

    fun translation(id: TranslationID): Translation? =
        all.firstOrNull { it.id == id }

    fun isBundled(id: TranslationID): Boolean =
        translation(id)?.isBundled ?: false
}
